sermant-plugins/sermant-mq-grayscale/sermant-mq-grayscale-plugin/src/main/java/io/sermant/mq/grayscale/utils/SubscriptionDataUtils.java

```diff
@@ -0,0 +1,418 @@
+            builder.append("'").append(tag).append("'");
+            builder.append(",");
+        }
+        builder.deleteCharAt(builder.length() - 1);
+        builder.append(RIGHT_BRACKET);
+        return builder.toString();
+    }
+
+    /**
+     * add current gray tags to sql92 expression
+     *
+     * @param originSubData originSubData
+     * @param addrTopicGroupKey addrTopicGroupKey
+     * @return sql expression
+     */
+    public static String addGrayTagsToSql92Expression(String originSubData, String addrTopicGroupKey) {
+        String originSubDataBak = originSubData;
+        if (!StringUtils.isBlank(originSubDataBak)) {
+            originSubDataBak = rebuildWithoutGrayTagSubData(originSubDataBak);
+        }
+        String sql92Expression = buildSql92Expression(addrTopicGroupKey);
+        if (StringUtils.isBlank(sql92Expression)) {
+            return originSubDataBak;
+        }
+        return StringUtils.isBlank(originSubDataBak)
+                ? sql92Expression : originSubDataBak + AND_SPLICE_STR + sql92Expression;
+    }
+
+    private static String buildSql92Expression(String addrTopicGroupKey) {
+        StringBuilder sb = new StringBuilder();
+        if (StringUtils.isEmpty(MqGrayscaleConfigUtils.getGrayGroupTag())) {
+            // base model return without exclude group message
+            if (CONSUME_TYPE_BASE.equals(MqGrayscaleConfigUtils.getConsumeType())) {
+                List<GrayTagItem> items = MqGrayscaleConfigUtils.getGrayTagItemByExcludeGroupTags();
+                if (!items.isEmpty()) {
+                    sb.append(buildBaseConsumerSql92Expression(items));
+                }
+                return sb.toString();
+            }
+
+            // auto model return without exclude group and current consume message gray group message
+            sb.append(buildBaseConsumerSql92Expression(getAutoTypeGrayTagItems(addrTopicGroupKey)));
+        } else {
+            MqGrayscaleConfig mqGrayscaleConfig = MqGrayscaleConfigUtils.getGrayscaleConfigs();
+            Optional<GrayTagItem> grayTagItem
+                    = mqGrayscaleConfig.getGrayTagByGroupTag(MqGrayscaleConfigUtils.getGrayGroupTag());
+            if (grayTagItem.isPresent()) {
+                sb.append(buildGrayConsumerSql92Expression(grayTagItem.get()));
+            } else {
+                LOGGER.warning(String.format(Locale.ENGLISH, "current gray group [%s] had not set grayscale, set it "
+                        + "and restart service to valid.", MqGrayscaleConfigUtils.getGrayGroupTag()));
+            }
+        }
+        return sb.toString();
+    }
+
+    private static List<GrayTagItem> getAutoTypeGrayTagItems(String addrTopicGroupKey) {
+        List<GrayTagItem> excludeItems = MqGrayscaleConfigUtils.getGrayTagItemByExcludeGroupTags();
+        List<GrayTagItem> autoDiscoveryGrayTags = AUTO_CHECK_GRAY_TAGS.get(addrTopicGroupKey);
+        if (autoDiscoveryGrayTags != null && !autoDiscoveryGrayTags.isEmpty()) {
```

Comment:
It is recommended to use CollectionUtils.isEmpty method

Reply:
if autoDiscoveryGrayTags==null it will throw exception